hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -188,6 +188,62 @@ object LowerTableIR {
           val env: Env[IR] = Env("row" -> row, "global" -> loweredChild.globals)
           loweredChild.mapPartition(rows => StreamFilter(rows, row.name, Subst(cond, BindingEnv(env))))
 
+        case TableHead(child, targetNumRows) =>
+          val loweredChild = lower(child)
+          val partitionSizeArray = loweredChild.mapPartition(rows => ArrayLen(ToArray(rows))).collect()
```

Comment:
I think the right solution here actually has a CDA inside the Recur. This will do a full scan of all the data up front -- what if you only need to look at 1 partition?

Reply:
Does this make sense? I can elaborate more tomorrow.

Reply:
I'll think about it for a bit and let you know if I have questions tomorrow. I see the problem, but not the solution yet.